llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2342,6 +2343,174 @@ SmallVector<utils::IteratorType> UnPackOp::getLoopIteratorTypes() {
+
+LogicalResult WinogradInputTransformOp::verify() {
+  Operation *op = getOperation();
+  if (getNumInputs() != 1) {
+    return op->emitOpError("expected one input operand");
+  }
+  if (getNumOutputs() != 1) {
+    return op->emitOpError("expected one output operand");
+  }
+  if (!input().getType().isa<ShapedType>()) {
+    return op->emitOpError("expected input type to be shaped");
+  }
+  if (!output().getType().isa<ShapedType>()) {
+    return op->emitOpError("expected output type to be shaped");
+  }
+  auto inputType = input().getType().cast<ShapedType>();
+  auto outputType = output().getType().cast<ShapedType>();
+  ArrayRef<int64_t> inputShape = inputType.getShape();
+  if (inputShape.size() != 4) {
+    return op->emitOpError("expected input operand to have rank 4");
+  }
+  ArrayRef<int64_t> outputShape = outputType.getShape();
+  if (outputType.getElementType() != inputType.getElementType()) {
+    return op->emitOpError(
+        "expected input/output element types to be identical");
+  }
+  if (getOutputOperandRank() != getInputOperandRank() + 2) {
+    return op->emitOpError(
+        "expected output rank to be equal to input rank + 2");
+  }
+  // Check shapes (assumes incoming input shape in NHWC)
+  const int64_t outputTileSize = getOutputTileSize();
+  const int64_t kernelSize = getKernelSize();
+  const int64_t inputTileSize = getInputTileSize();
+  SmallVector<int64_t> expectedOutputShape(getInputOperandRank() + 2,
+                                           inputTileSize);
+  for (int i = 0; i < inputShape.size(); i++) {
+    if ((i == 0) || (i == 3) || (inputShape[i] == ShapedType::kDynamicSize)) {
+      expectedOutputShape[i + 2] = inputShape[i];
+    } else {
+      expectedOutputShape[i + 2] =
+          std::ceil((float)(inputShape[i] - kernelSize + 1) / outputTileSize);
+    }
+  }
+  if (llvm::any_of(llvm::zip(expectedOutputShape, outputShape),
+                   [](std::tuple<int64_t, int64_t> s) {
+                     return std::get<0>(s) != ShapedType::kDynamicSize &&
+                            std::get<1>(s) != ShapedType::kDynamicSize &&
+                            std::get<0>(s) != std::get<1>(s);
+                   })) {
+    return op->emitOpError("incompatible output shape");
+  }
+  const StringRef tensorFormat = getTensorFormat();
+  if (tensorFormat != "nhwc") {
+    return op->emitOpError("only nhwc format currently supported");
+  }
+  return success();
+}
+
+SmallVector<Range>
+WinogradInputTransformOp::getIterationDomain(OpBuilder &builder) {
+  int64_t operandRank = getIterationDomainRank();
+  SmallVector<Range> loopBounds(operandRank);
+  Location loc = getLoc();
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  Value source = input();
+  // Expose only the batch dimension(0) and channel dimension(3) for NHWC tensor
+  SmallVector<int64_t> indices = {0, 3};
```

Comment:
I see that you have `getIterationDomainRank` to abstract away the dimensions exposed to tiling interface, but now this is hard coded. I am guessing if you use `image_dimensions`, then all dimensions that are not image dimensions will be used here.